Hostile mobs

someaddons opened the pull request:
# Changes proposed in this pull request:
## Features:
- adding config option to make mobs hostile to citizens
- adding config option to let citizens and guards call for help when attacked
- citizens run away after getting hit and keeps avoiding the attacker(if there is one)
- reworking old avoidance AI, now only scans the area actively after beeing attacked before.
- adding mercenaries which defend the colony for a price
    - summoned at the townhall via a button
    - last 20 min and reappear to be hired after 40min(so a 20min cooldown)
- adding housing to guards in guardtowers


## Fixes:
- fix child sounds not beeing registered
- fix damage sources of our entites to be entitydamagesources (so attacked citizens recognise raiders correctly)
- fix raider attackspeed, previous formula was broken and not scaling at all
- fix raiders stuck recognition causing them to ignore entity collision
- fixing barracks citizen limit recalculation, was only recalculated on building upgrade instead of citizen assignment
- fixed two pathing bugs which caused entites to not move towards their target frequently, e.g. some raiders not moving to the colony.
- fix lumberjack trying to save null values to nbt

Review please

someaddons committed:
- adding config option to make mobs hostile to citizens
- adding config option to let citizens call for help when attacked
- reworking old avoidance AI, now only scans the area actively after beeing attacked before.

someaddons committed:
codacy fixes

OCDCleric commented:
I feel like if this gets implemented that it should be default: false or have 1 or 2 guards spawn in free for early game safety

someaddons commented:
I think it is natural in our progression that at the start your citizens are vulnerable and require caretaking

OCDCleric commented:
True, but this also means the prepatory period is probably going to be much longer. Worst case is that we start putting up walls around town hall first thing.

someaddons commented:
Raiders happen realtively early aswell and are not preventable by lighting up the area, so should probably have both on the same setting

OCDCleric commented:
Will the mobs trigger mourning if they kill citizens?

Raycoms commented:
I would make it default false for the beginning and then turn it on after
we got some feedback on it.

Am So., 16. Juni 2019 um 12:56 Uhr schrieb someaddons <
notifications@github.com>:

> Raiders happen realtively early aswell and are not preventable by lighting
> up the area, so should probably have both on the same setting
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/3673?email_source=notifications&email_token=ABRD3SZW5N7LDPJDKC575MLP2ZPCZA5CNFSM4HYP24I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXZP46A#issuecomment-502464120>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRD3S554CWQS3DYOCACF3LP2ZPCZANCNFSM4HYP24IQ>
> .
>

someaddons commented:
ye thats fine, we can also increase early defenses by e.g. giving the guard tower an additional guard, or letting it house the citizen

OCDCleric commented:
Actually couldn't we have citizens fight? Such as;

1. If attacked search for Guards
 -If Yes, run away and call help
 -If No, Punch or Use Tools on hand to fight back.
2. If attacked search for Guards
 -If Yes, run away call for help
 -If No, call for help from nearby citizens and gang up on enemy mob.

marchermans commented:
I think at least minimal self defence of citizens needs to be possible.

someaddons commented:
ye I could add a small AI for the to fight back, want them to fight always, or just when no guards are close?

Raycoms commented:
I'd say only when no guards are close, maybe even say that citizens within
a range come to help. Reminds me of aoe

On Sun, Jun 16, 2019, 16:30 someaddons <notifications@github.com> wrote:

> ye I could add a small AI for the to fight back, want them to fight
> always, or just when no guards are close?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/3673?email_source=notifications&email_token=ABRD3S2WO4VIFRU7GVZLXXLP22IETA5CNFSM4HYP24I2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXZTTBA#issuecomment-502479236>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRD3S736UHML4PSJDGMA2DP22IETANCNFSM4HYP24IQ>
> .
>

ArkenorOakshadow commented:
Could vary by class. The lumberjack has an axe, after all, so is better situated to put up a fight than, say, the baker.

OCDCleric commented:
At worst could add a mercenary contract to town hall that creates patrolling mercs for a period of time?

marchermans commented:
@OCDCleric That is a nice Idea.
@someaddons Could you take a look at that as well, spawn in temporary guards that can be bought in the TownHall?

someaddons commented:
ye I could, would be kinda similar to raiders then. What kinda cost do you think one could use?

someaddons commented:
So to sum it up:
- have mobs attack citizen
- citizen call guards and flee
- citizen fight back when no guards, calling nearby citizen
- guardtowers provide housing to guards as barracks
- Townhall option to spawn temporary protector guards

marchermans commented:
Correct 😄  👍 

[![Approved with PullApprove](https://img.shields.io/badge/maintainers-approved-brightgreen.svg)](https://pullapprove.com/ldtteam/minecolonies/pull-request/3673/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

marchermans commented:
Balance it a bit. I think a base cost, plus some cost per minecraft day?

someaddons commented:
not sure which resource one could use as costs, my ideas would be:
- reduction in happiness when hiring the mercenaries
- mercenaries taking 70% of all new resources produced
- a direct material cost consumed when hiring, e.g. 100 wood
- a gold currency which we dont have yet :D

marchermans commented:
Yeah, I was thinking something in that direction as well.

OCDCleric commented:
1. Reductuon in happiness makes sense
2. What woul 70% of all new resources entail?
3. Material cost would be Food and Gold?
4. I sense a new worker for making currency.

Ps. Should probably make a different thread elsewhere to didcuss this lol.

someaddons committed:
- adding mercenaries which defend the colony for a price
- fixing barracks citizen limit recalculation
- adding housing to guards in guardtowers
- fixed two pathing bugs which caused entites to not fully move towards their target frequently.

someaddons committed:
codacy fixes

someaddons committed:
codacy round two

someaddons committed:
codacy v3

someaddons committed:
codacy v4

someaddons committed:
codacy v5

someaddons committed:
- fix child sounds
- fix damage sources
- fix raider attackspeed

someaddons committed:
- fix raiders stuck recognition

someaddons committed:
- recalc max citizens also on guard unassign

someaddons committed:
- adding missing final

someaddons committed:
fix name

someaddons committed:
Merge remote-tracking branch 'remotes/upstream/version/1.12' into MobAI

Raycoms committed:
Merge branch 'version/1.12' into mobAI

Raycoms commented on src/api/java/com/minecolonies/api/configuration/Configurations.java:
```diff
@@ -113,6 +113,12 @@
         public int minimumNumberOfNightsBetweenRaids = 1;
 
+        @Config.Comment("Should Mobs attack citizens? default:false")
+        public boolean mobAttackCitizens = true;
```
default false, but here it is true? =D

someaddons replied:
oh forgot to change it back from the testing version :D

Raycoms commented on src/api/java/com/minecolonies/api/entity/ai/pathfinding/AbstractWalkToProxy.java:
```diff
@@ -102,7 +102,7 @@ public boolean walkToBlock(@NotNull final BlockPos target, final int range, fina
 
-            this.resetProxyList();
+            proxyList.clear();
             return takeTheDirectPath(target, range, onMove);
```
And I thought both things do the same thing =P

someaddons replied:
yea there is a subtle difference, which is that resetProxyList() also clears the currentProxy which just was set on the line before this

Raycoms commented on src/api/java/com/minecolonies/api/util/FireworkUtils.java:
```diff
@@ -22,15 +22,16 @@
      */
-    private FireworkUtils() {
+    private FireworkUtils()
+    {
```
ty =D

Raycoms commented on src/main/java/com/minecolonies/coremod/client/render/mobs/RenderMercenary.java:
```diff
@@ -0,0 +1,35 @@
+     */
+    private static final ResourceLocation TEXTURE = new ResourceLocation("minecolonies:textures/entity/settlermale1.png");
+
+
```
seems to be missing a bit of doc here

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/Colony.java:
```diff
@@ -1563,4 +1572,31 @@ public void increaseBoughtCitizenCost()
+    public void registerEntityWithColony(final IColonyRelatedEntity entity)
+    {
+        // Add entity registry logic here
+    }
```
is there something missing here?

someaddons replied:
oh thats just unused atm, not sure if we got a use for it yet. Probably should just remove it so we dont have new unused code

Raycoms replied:
Agreed

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/ColonyView.java:
```diff
@@ -992,4 +1000,14 @@ public boolean isRaiding()
         return this.horde > 0;
     }
+
+    public long getMercenaryLastUseTime()
```
@Override

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java:
```diff
@@ -364,6 +385,81 @@ public boolean attackEntityFrom(@NotNull final DamageSource damageSource, final
+        // Makes the avoidance AI take over.
+        currentlyFleeing = true;
+
+        if ((getCitizenJobHandler().getColonyJob() instanceof AbstractJobGuard))
```
Should the guards also move away?

someaddons replied:
guards wont move away, as they do not have the moveawayAI added at all.

Raycoms replied:
moveAwayPath = this.getNavigator().moveAwayFromEntityLiving(this, 5, INITIAL_RUN_SPEED_AVOID)

but it does this for the guards, isn't that unnecessary overhead?

someaddons replied:
no this is intended, it is running away a little bit from the current pos when getting hurt by environmental, nonentity caused damage

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/mobs/EntityMercenary.java:
```diff
@@ -0,0 +1,518 @@
+     * Proxy for cheaper pathing.
+     */
+    private              GeneralEntityWalkToProxy proxy;
+    /**
```
this formatting seems off.

Raycoms commented:
Ahh okay

Am Sa., 13. Juli 2019 um 14:31 Uhr schrieb someaddons <
notifications@github.com>:

> *@someaddons* commented on this pull request.
> ------------------------------
>
> In src/main/java/com/minecolonies/coremod/entity/EntityCitizen.java
> <https://github.com/ldtteam/minecolonies/pull/3673#discussion_r303212743>:
>
> > +     */
> +    private void performMoveAway(@Nullable final Entity attacker)
> +    {
> +        this.getCitizenStatusHandler().setLatestStatus(new TextComponentTranslation("com.minecolonies.coremod.status.avoiding"));
> +
> +        // Environmental damage
> +        if (!(attacker instanceof EntityLivingBase))
> +        {
> +            moveAwayPath = this.getNavigator().moveAwayFromEntityLiving(this, 5, INITIAL_RUN_SPEED_AVOID);
> +            return;
> +        }
> +
> +        // Makes the avoidance AI take over.
> +        currentlyFleeing = true;
> +
> +        if ((getCitizenJobHandler().getColonyJob() instanceof AbstractJobGuard))
>
> no this is intended, it is running away a little bit from the current pos
> when getting hurt by environmental, nonentity caused damage
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/3673?email_source=notifications&email_token=ABRD3SZVYUSGJ3LUXCNGZFLP7IGP3A5CNFSM4HYP24I2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6LO7SQ#discussion_r303212743>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABRD3S5KECO3TI4EUMDH4U3P7IGP3ANCNFSM4HYP24IQ>
> .
>

Raycoms committed:
Merge branch 'version/1.12' into mobAI

someaddons committed:
small adjustments

someaddons committed:
Merge remote-tracking branch 'remotes/upstream/version/1.12' into MobAI

someaddons committed:
Merge remote-tracking branch 'remotes/origin/mobAI' into MobAI

someaddons committed:
merging defaults & armor commits
